Fixes multiple errors within the Mixed Models analysis

FBartos opened the pull request:
Fixes https://github.com/jasp-stats/jasp-test-release/issues/862
Fixes https://github.com/jasp-stats/jasp-test-release/issues/863
Fixes https://github.com/jasp-stats/jasp-test-release/issues/864
Fixes https://github.com/jasp-stats/jasp-test-release/issues/865
Fixes https://github.com/jasp-stats/jasp-test-release/issues/866
Fixes https://github.com/jasp-stats/jasp-test-release/issues/892
Fixes https://github.com/jasp-stats/jasp-test-release/issues/893

Fixes https://github.com/jasp-stats/jasp-test-release/issues/875
Fixes https://github.com/jasp-stats/jasp-test-release/issues/878
Fixes https://github.com/jasp-stats/jasp-test-release/issues/879
Fixes https://github.com/jasp-stats/jasp-test-release/issues/881
Fixes https://github.com/jasp-stats/jasp-test-release/issues/895

FBartos committed:
fixes: #862

A fitted model was deleted after setting an error withing the summary table. However, next ioption refreshed the model and it wasn't reset to null since the summary table wasn't called - the other output options could've been accessed without a proper model.
Fixed by creating a state that checks whether the model contained error. Doesn't allow to proceed unless a proper model was fitted.

FBartos committed:
only GLMM didn't allow "nominalText" as a fixed effects

FBartos committed:
fixes: #863

Only for LMMs require emmeans a type of df calculation to be specifie. If a GLMM model with gaussian family and identity link function is fitted, emmeans treat it as an LMM...
The same error would've occured for trends too, fixed both.

FBartos committed:
fixes: #864

The variable in "Number of trials" was still being passed resulting in the error. I dealt with the situatuation in the same way as Bruno did in (same way as in https://github.com/jasp-stats/jasp-desktop/pull/4142). I hide and remove the variable from "Number of trials" if family is different that "Binomial ("Aggregated"). It should hopefully remove the arrow as well when Bruno merges his changes to the AssignedVariablesList.

vandenman commented:
@FBartos if you start your message with "Fixes \<link to issue\>" then the linked issues will appear on the right under linked issues and can be automatically closed when merging this PR.

vandenman commented:
So all the errors linked in the issues are fixed, great! 

I did notice one more issue though:
![image](https://user-images.githubusercontent.com/21319932/85513963-d00aaa00-b5fb-11ea-9867-446e315ffe6d.png)

The variable names (e.g., `JaspColumn_.15._Encoded`) are not decoded before they are shown in the error message. To reproduce:

1. `debString` & `contGamma` to fixed effects variables.
2. `conBinom` as random effects grouping factors.
3. Family is `Gaussian`.
4. dependent is `contNormal`.

FBartos committed:
fixes #875

If R-hat is Inf, the max function returns NaN and it fails the logical test. A similar problem could've happend with ESS, which is produces as NaN directly by rstan

FBartos committed:
translating variable names within an error message

FBartos commented:
thanks Don, I will be adding the links now.

also, fixing this issue
![image](https://user-images.githubusercontent.com/38475991/85524170-2e3d8a00-b608-11ea-9061-935f454bbfbf.png)

FBartos committed:
fixes #878

FBartos committed:
fever -> fewer

vandenman commented:
@FBartos let me know when this is ready to review!

FBartos committed:
fixes #881

FBartos committed:
fixes #865

by adding a checks for collinearity

FBartos committed:
fixing more typos (Frantisek's English sucks)

FBartos committed:
fixes #895

afex_plot requires 'dv' to be supplied when using stanova model

FBartos committed:
redoing 879 after rebase

FBartos committed:
fixes 892

by increasing minumum boostrap samples to 100

FBartos committed:
fixing 893 (and another Frantisek's spelling mistake, his grammar sucks vol2)

FBartos commented:
I think that this should fix all that was fixable within https://github.com/jasp-stats/jasp-test-release/issues/768, https://github.com/jasp-stats/jasp-test-release/issues/769, https://github.com/jasp-stats/jasp-test-release/issues/770, and https://github.com/jasp-stats/jasp-test-release/issues/771 as of 24. 06 (20:45). Could you review it again please? @vandenman

vandenman commented on JASP-Engine/JASP/R/MixedModelsCommon.R:
```diff
@@ -491,18 +492,27 @@
-        jaspResults[["ANOVAsummary"]] <- ANOVAsummary
-        jaspResults[["mmModel"]]      <- NULL
+        
+        hasERROR <- createJaspState()
+        hasERROR$object <- TRUE
+        hasERROR$dependOn(c(dependencies, seed_dependencies))
+        jaspResults[["hasERROR"]] <- hasERROR
```
Do we need this? Can't we just do `ANOVAsummary$dependOn(c(dependencies, seed_dependencies))` before the return and `ANOVAsummary$dependOn(c(dependencies, seed_dependencies, "pvalVS"))` after the if statement?

FBartos replied:
It's a little akward but I cannot set the `jaspResults[["mmModel"]]` to `NULL` in order to skip the block with summary and plotting functions because the model would be reffited, and it would skip the main summary that was already produced (inside of which it's set to NULL, I cannot do that before since I need to keep the error message). This object keeps track on whether the model failed and allows to skip the summary and plotting functions. I could alternatively create a designated function that would set the model to null after creating the summary table and it wouldn't be dependent on it. It might be a better solution.

FBartos replied:
however, it would resulted on refiting the failed model every time, even when requested

vandenman replied:
> This object keeps track on whether the model failed

Okay, but if you set an error on any jaspResults object you can check for an error using `$getError()`.

FBartos replied:
oh, great. That's much better solution. Thanks

vandenman commented on JASP-Engine/JASP/R/MixedModelsCommon.R:
```diff
@@ -2877,6 +2906,18 @@
+   diag(cor_mat) <- 0
+   cor_mat[lower.tri(cor_mat)] <- 0
+   if(any(1 - abs(cor_mat) < 1e-5)){
+     var_ind   <- which(abs(cor_mat) == 1, arr.ind = TRUE)
```
```suggestion
   nearOne <- 1 - abs(cor_mat) < sqrt(.Machine$double.eps)
   if(any(nearOne)){
     var_ind   <- which(nearOne, arr.ind = TRUE)
```
The issue is that you check with a tolerance in the if statement but without a tolerance in `which`. 

Using `sqrt(.Machine$double.eps)` (~1.490116e-08) should make this less machine specific, feel free to use a different root though (e.g., `.Machine$double.eps^(1/3)`).

FBartos committed:
one last renaming that was lost during the rebase

vandenman reviewed (changes requested):
Two errors remain, the others seem fixed.

Error message for equal variables doesn't use the variable names (see comment in the code):
![image](https://user-images.githubusercontent.com/21319932/85673638-62768080-b6c4-11ea-9b22-d4a18051fe05.png)


There's still an error if warmup == iterations:

![image](https://user-images.githubusercontent.com/21319932/85673543-4672df00-b6c4-11ea-9d3f-86e8fc1041a5.png)

FBartos committed:
machine tollerance

Co-authored-by: Don van den Bergh <donvdbergh@hotmail.com>

FBartos committed:
changing the +2 to the correct value

FBartos committed:
removing the "hasERROR" jasp state

FBartos commented:
Thanks for the review Don. I found that I mistakenly changed the default value instead of the min value after I rebased on the updated qml interface. That's fixed now. I incorporated the suggested machine tolerance and restructured the check for errors as suggested.

vandenman commented:
Changes to the code look fine!

Two more bugs:

#### 1
![image](https://user-images.githubusercontent.com/21319932/85694940-e0905280-b6d7-11ea-9b60-c6485bc9d14e.png)

1. contBinom to fixed
2. facGender to Random effects
3. `Adapt delta` to 0.0
4. contNormal to dependent.

perhaps `Adapt delta` needs a larger minimum.


#### 2
You can't set `Iterations` below `Warmup + 2`, but it does work the other way around.

Set:
1. `Warmup` = 1
2. `Iterations` = 3
3. `Warmup` = 100

The interface adds a red box around `Iterations` but still the analysis crashes. Error message was the same as in the screenshot above.

FBartos commented:
I'm not sure whether we can prevent (2) from withing the qml. Are you OK if I add an error check next to the datachecks and exit the analysis?

FBartos committed:
always eval cont.bayes (should help with #884 and similar)

FBartos committed:
error handling for B(G)LMM fits

FBartos committed:
adding special check for iterations vs warmup

FBartos committed:
more reasobable min values for the B(G)LMM

FBartos commented:
Also should
fix https://github.com/jasp-stats/jasp-test-release/issues/884
but it's not really testable.

That should be all the changes
